### ISSUE TITLE
Connection over gateway doesn't work with OpenSSL 3.0

### ISSUE THREAD

**Issue by ondrejholy** [FreeRDP/FreeRDP#7449]
**Describe the bug**
Connection over gateway doesn't work currently with up-to-date `master`/`stable-2.0` branches when build with OpenSSL 3.0. Individual connections to both machines work properly. It also works with older OpenSSL versions. Any idea what might be wrong?

**To Reproduce**
1. Build freerdp with OpenSSL 3.0, or use e.g. `xfreerdp` from Fedora rawhide:
1. `xfreerdp /v:IP /g:GATEWAYIP /u:USER /p:PASS /redirect-prefer:ip`
2. It fails with:
```
[11:41:14:369] [335533:335534] [ERROR][com.freerdp.core.gateway.http] - http_response_recv: Retries exceeded
[11:41:14:369] [335533:335534] [ERROR][com.freerdp.core.gateway.http] - 404629DD7E7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:346:Global default library context, Algorithm (MD4 : 84), Properties ()
[11:41:14:369] [335533:335534] [ERROR][com.freerdp.core.gateway.http] - 404629DD7E7F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:234:
[11:41:14:369] [335533:335534] [ERROR][com.freerdp.core.gateway.http] - 404629DD7E7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:346:Global default library context, Algorithm (MD4 : 84), Properties ()
[11:41:14:369] [335533:335534] [ERROR][com.freerdp.core.gateway.http] - 404629DD7E7F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:234:
[11:41:14:370] [335533:335534] [ERROR][com.freerdp.core.nego] - Protocol Security Negotiation Failure
[11:41:14:370] [335533:335534] [ERROR][com.freerdp.core] - rdp_client_connect:freerdp_set_last_error_ex ERRCONNECT_SECURITY_NEGO_CONNECT_FAILED [0x0002000C]
[11:41:14:370] [335533:335534] [ERROR][com.freerdp.core.connection] - Error: protocol security negotiation or connection failure
```
3. With #7448 applied on the top, the error changed, but doesn't work anyway:
```
[11:44:12:465] [336396:336397] [ERROR][com.freerdp.core] - rdg_process_close_packet:freerdp_set_last_error_ex E_PROXY_INTERNALERROR [0x800759D8]
```

**Expected behavior**
FreeRDP connects successfully to `IP` over `GATEWAYIP`

**Application details**
* FreeRDP version (`xfreerdp /version`)
* Command line used
* Output of `xfreerdp /buildconfig`
* OS version connecting to (server side)
Windows Server 2019
* If available the log output from a run with `/log-level:trace 2>&1 | tee log.txt` 
There is a log on top of #7448: [log.txt](https://github.com/FreeRDP/FreeRDP/files/7562635/log.txt).
* If you built it yourself add some notes which tag/commit/branch you have used, also your cmake parameters and
  compiler can help

**Environment:**
 - OS: Linux
 - Version/Distribution: Fedora 35
 - Architecture: amd64

**Additional context**
~~It works fine when hostname is used instead of `IP` for `/v:` with https://github.com/FreeRDP/FreeRDP/pull/7448. This might determine the potentially problematic code...~~

There is a backtrace for `freerdp_set_last_error_ex` on top of #7448: [backtrace.txt](https://github.com/FreeRDP/FreeRDP/files/7562636/backtrace.txt).

**Comment by akallabeth** [comment]
@ondrejholy no idea.
as for the log on top of #7448 , it looks like only `stdout` was redirected? there is no `stderr` message in there

**Comment by akallabeth** [comment]
@ondrejholy just to be sure, if you skip `/redirect-prefer:ip` does it work then? the log message `E_PROXY_INTERNALERROR` seems to originate at the gateway side?

**Comment by ondrejholy** [comment]
@akallabeth The `stderr` output is also included, there are just two `[WARN]` lines and the `E_PROXY_INTERNALERROR` line almost at the beginning of the log.

It fails the same way without `/redirect-prefer:ip` with OpenSSL 3.0. (I have been convinced that `/redirect-prefer:ip` has to be used when using IP addresses because it didn't work earlier for me with OpenSSL 1.1, but it seems that it works now with 1.1, hmm.)

Yes, it seems that the `E_PROXY_INTERNALERROR` error is a gateway reply, but it must be something with the client, when it works with older OpenSSL versions...

**Comment by akallabeth** [comment]
@ondrejholy ok, missed these (expected that at the end of the file)
so, if I read this correctly:

1. HTTP gateway, transport upgraded to websockets
2. The server side does not like something, even before authentication

things that come to mind are:
* `TLS` requirements of client too strict? (e.g. `RDP` usually lags behind in enforcing newer revisions)

**Comment by ondrejholy** [comment]
@akallabeth It is true, that OpenSSL manual mentions possible TLS hand-share failures: https://wiki.openssl.org/index.php/OpenSSL_3.0#SSL.2FTLS, but https://github.com/FreeRDP/FreeRDP/pull/7448 should load all the necessary providers. Not sure how to verify the `TLS` requirements, but I suppose that individual connections would not work in that case as well, but they work, so this must be something more gateway-specific...

**Comment by akallabeth** [comment]
@ondrejholy Gateway does not connect to the RDP server but some web server (IIS on windows, for the log you posted it also uses websockets)
maybe that is the key difference?

**Comment by jay832473** [comment]
Hello,

I'm using freerdp 2.4.1 and the only thing which seems missing to support openssl 3.0.0 is to add

OSSL_PROVIDER_load(NULL, "legacy");
OSSL_PROVIDER_load(NULL, "default");

in function tls_prepare() in file libfreerdp/crypto/tls.c 
before SSL_CTX_new() is called

Then openssl will support obsolete/unsecure algorithms also
You might also want to add support for KTLS with SSL_OP_ENABLE_KTLS

**Comment by ondrejholy** [comment]
@jay832473 This should be fixed in the next version thanks to https://github.com/FreeRDP/FreeRDP/pull/7448. Although it fixes various issues, this gateway issue still remains...

**Comment by mpharoah** [comment]
Still broken on master when using a gateway. I was able to connect by building an xfreerdp flatpak in a runtime with an older version of OpenSSL.

Here's instructions on how to do that for anyone who is locked out of being able to remote desktop into things due to this:

1. Install `flatpak-builder` and `flatpak`
2. Create a file named `com.freerdp.FreeRDP.json`
3. Paste the following into said file:
```json
{
	"id": "com.freerdp.FreeRDP",
	"runtime": "org.freedesktop.Platform",
	"runtime-version": "20.08",
	"sdk": "org.freedesktop.Sdk",
	"command": "xfreerdp",
	"finish-args" : [
		"--socket=x11",
		"--socket=wayland",
		"--socket=pulseaudio",
		"--share=ipc",
		"--share=network"
	],
	"add-extensions": {
		"org.freedesktop.Platform.ffmpeg-full": {
			"directory": "lib/ffmpeg",
			"version": "20.08",
			"add-ld-path": "."
		}
	},
	"modules": [
		{
			"name": "xfreerdp",
			"buildsystem": "cmake",
			"config-opts": [
				"-DCHANNEL_URBDRC=OFF",
				"-DWITH_OPENH264:BOOL=ON"
			],
			"sources": [{
				"type": "git",
				"url": "https://github.com/FreeRDP/FreeRDP.git",
				"commit": "c3df0be63953ed98525d9b736ba878ad733de059"
			}]
		}
	]
}

```
4. Run the following commands to build and install the flatpak:
```bash
flatpak-builder build --install-deps-from=flathub --force-clean --ccache com.freerdp.FreeRDP.json
flatpak build-export repo build
flatpak remote-add --no-gpg-verify --if-not-exists xfreerdp-local repo
flatpak install xfreerdp-local com.freerdp.FreeRDP
```
5. You should now be able to run `xfreerdp` using `flatpak run com.freerdp.FreeRDP` followed by your usual command line arguments

**Comment by omriasta** [comment]
This worked for me! Currently using Remmina and noticed this as soon as I updated to Ubuntu 22.04. Is there anyway to fix this so I can use Remmina? I'm guessing the plugin needs to be updated somehow?

**Comment by akarl10** [comment]
Since ubuntu 22.04 uses openssl3 I am bugged by this issue too.
If allowed, /sec:rdp does indeed work over rdg, but tls based do not.
What I observed was that somehow openssl 3 does omit the outermost header. A tls handshake usually stars with
0x16 0x03 0x01 b1 b2 (Hanshake, ssl 3.1, length), but this 5 bytes are missing in the inner ssl connection.
Does someone know the inners of openssl and understand how this can happen?
![grafik](https://user-images.githubusercontent.com/3184745/164981395-29e9e849-98b4-4b4a-a0a7-72fa28055ead.png)
![grafik](https://user-images.githubusercontent.com/3184745/164981505-1cd5c22b-ff65-4511-bd6c-979ba5b7c1eb.png)

**Comment by akarl10** [comment]
Ok, a least on my system I got it working (tested only with rdg, not rpc, but the issue is most likly the same).

I will create a push request when I have done some cleanup and documentation

**Comment by akallabeth** [comment]
Does work with our test setups here, so closing.

**Comment by ondrejholy** [comment]
It works in our test setup as well, thanks!

**Comment by Jcs1994** [comment]
I don't get it.. when is this issue closed ? how can i fix this issue ?

**Comment by Jcs1994** [comment]
Oh, the fix is in the 2.7.0 but my version is still 2.6.1
when will the update will be available via apt ?
